File: typescript/examples/getting-started/04c-edit-image.ts
// Image editing — pass a reference image to generateImage() to transform it.
// The platform routes to image_gen() with the reference image included.
import { readFileSync, writeFileSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { z } from "zod";
import { Opper } from "../../src/index.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const opper = new Opper();

// ── Option 1: Schema-driven call() ──────────────────────────────────────────
// Schemas alone tell the platform this is an image edit — include a reference_image field.

const imageBase64 = readFileSync(resolve(__dirname, "media/image.png")).toString("base64");

const raw = await opper.call("sdk-test-edit-image-raw", {
  input_schema: z.object({
    description: z.string().describe("Text description of the image to generate"),
    reference_image: z.string().describe("Base64-encoded reference image"),
  }),
  output_schema: z.object({
    image: z.string().describe("Base64-encoded image data"),
    mime_type: z.string().describe("MIME type of the generated image"),
  }),
  input: {
    description: "Give the cat some sunglasses and a party hat",
    reference_image: imageBase64,
  },
});

console.log("── Schema-driven call() ──");
console.log("MIME type:", raw.data.mime_type);
console.log("Image base64 length:", raw.data.image.length);

const ext = raw.data.mime_type.split("/")[1] || "png";
const rawPath = resolve(__dirname, `media/edited-image-raw.${ext}`);
writeFileSync(rawPath, Buffer.from(raw.data.image, "base64"));
console.log(`Saved to ${rawPath}`);

// ── Option 2: Convenience method ────────────────────────────────────────────
// Accepts a file path directly — the SDK reads and base64-encodes it for you.

const easy = await opper.generateImage("sdk-test-edit-image", {
  prompt: "Give the cat some sunglasses and a party hat",
  reference_image: { path: resolve(__dirname, "media/image.png") },
});

console.log("\n── Convenience method ──");
console.log("MIME type:", easy.data.mime_type);
console.log("Image base64 length:", easy.data.image.length);

const easyPath = easy.save(resolve(__dirname, "media/edited-image"));
console.log(`Saved to ${easyPath}`);

File: python/examples/getting-started/04c_edit_image.py
# Image editing — pass a reference image to generate_image() to transform it.
# The platform routes to image_gen() with the reference image included.
import base64
from pathlib import Path

from opperai import Opper

media_dir = Path(__file__).parent / "media"
opper = Opper()

# ── Option 1: Convenience method ────────────────────────────────────────────
# Accepts a file path directly — the SDK reads and base64-encodes it for you.

easy = opper.generate_image(
    "sdk-test-edit-image",
    prompt="Give the cat some sunglasses and a party hat",
    reference_image=str(media_dir / "image.png"),
)

print("── Convenience method ──")
print("Image data keys:", list(easy.data.keys()) if isinstance(easy.data, dict) else type(easy.data))
print("Meta:", easy.meta)

# ── Option 2: Raw call() with explicit schemas ─────────────────────────────

image_b64 = base64.b64encode((media_dir / "image.png").read_bytes()).decode()

raw = opper.call(
    "sdk-test-edit-image-raw",
    input={
        "description": "Give the cat some sunglasses and a party hat",
        "reference_image": image_b64,
    },
    input_schema={
        "type": "object",
        "properties": {
            "description": {"type": "string", "description": "Text description of the image to generate"},
            "reference_image": {"type": "string", "description": "Base64-encoded reference image"},
        },
    },
    output_schema={
        "type": "object",
        "properties": {
            "image": {"type": "string", "description": "Base64-encoded image data"},
            "mime_type": {"type": "string", "description": "MIME type of the generated image"},
        },
    },
)

print("\n── Raw call() ──")
print("MIME type:", raw.data.get("mime_type"))
image_data = raw.data.get("image", "")
print("Image base64 length:", len(image_data))

ext = (raw.data.get("mime_type", "image/png")).split("/")[-1]
raw_path = media_dir / f"edited-image-raw.{ext}"
raw_path.write_bytes(base64.b64decode(image_data))
print(f"Saved to {raw_path}")
